### ISSUE TITLE
[Feature request] - allow Texture Packer to place sprites at coordinates divisible by specifiable values

### ISSUE THREAD

**Issue by Adam13531** [libgdx/libgdx#5603]
### The request itself
The request is to expose options like [CodeAndWeb's TexturePacker](https://www.codeandweb.com/texturepacker)'s [`commonDivisorX` and `commonDivisorY`](https://www.codeandweb.com/texturepacker/documentation/custom-exporter). When specified, libgdx's texture packer would only place sprites at multiples of those coordinates.

For example, specifying both values as `16` would result in each sprite's final coordinates being divisible by 16.

### Rationale

This resulted from a PixiJS issue that I filed [here](https://github.com/pixijs/pixi.js/issues/5549). When mipmapping sprites from a spritesheet, the sprite's source coordinates affect whether the result will be blurry. I believe the texture packer should be able to place sprites in a way that this issue can be solved outside of application code completely.

### Proposed solution

I don't have the time/bandwidth right now to be able to test this out, but here's an algorithm I came up with:

- Before attempting to place a sprite, expand its bounding rectangle by `commonDivisorX` and `commonDivisorY`, then try to place the now-larger rectangle.
- After finding a position, adjust the final coordinates so that they're divisible by the divisors specified.
- Emit the final rectangle's size as its original size.

For example, suppose you're trying to place a rectangle of size 100x50, and both divisors are set to 16.

- First, we increase the size of the rectangle temporarily to accommodate the divisors, so its size becomes 116x66.
- Suppose the packer computes the coordinates to be (90, 30). We know that 16 pixels in either direction are the fake padding that we added, so we can nudge the rectangle over to (96, 32) so that both coordinates are divisible by 16.
- Finally, we set the rectangle's size back to the original size of 100x50, that way other rectangles can be packed more closely to the right and bottom sides of the one we just placed.

While I'm pretty sure it will work, I don't know if it would result in the best possible packing of sprites.

**Comment by Darkyenus** [comment]
You will always have problems with mipmapping on a texture atlas, if not on the first mipmap level, then on the smaller ones. That said, could you try enabling [`duplicatePadding`](https://github.com/libgdx/libgdx/wiki/Texture-packer) (and related options) to see if it helps?

**Comment by Adam13531** [comment]
> You will always have problems with mipmapping on a texture atlas, if not on the first mipmap level, then on the smaller ones.

@darkyenus Could you clarify this a bit? In my case, I know the smallest scale that I'm going to allow the sprite to render at—let's say it's 1/12 the original size—so my understanding is that the next lowest mip would be used, i.e. `nextPow2(12) == 16`. I thought that meant that my sprites would just need to be at offsets divisible by 16 in my atlas for them to render clearly at a scale no lower than 1/16 the original size. In testing this with a single sprite, it seemed to hold true, but to convert all of my sprites to coordinates divisible by 16 would not have to be automated.

I'll have to try out `duplicatePadding` if there's time this week. I was actually considering purchasing a license to the other TexturePacker software tomorrow, but if what you're saying is correct for my case, then it sounds like a texture packer *can't* solve my issue.

Also, thank you for the response; I appreciate the help!

**Comment by Darkyenus** [comment]
> Could you clarify this a bit?

Search for "mipmap texture atlas" and you'll find many people discussing many different problems with this setup. Both alignment and padding will probably help, but it will never be perfect/painless.

**Comment by Adam13531** [comment]
I had time to do some research on this!

I tried out `duplicatePadding` in libgdx's texture packer, but it doesn't control the resulting offsets of sprites in the spritesheet at all, so it didn't help me for my issue.

I tried out [TexturePacker](https://www.codeandweb.com/texturepacker) with a `commonDivisorX` and `commonDivisorY` of 16, and my sprites did render correctly as I'd hoped.

[I mentioned in the PixiJS issue](https://github.com/pixijs/pixi.js/issues/5549#issuecomment-478313396) that I could bypass this issue by rendering to a mipmappable texture in memory and then using that as the foundation for my sprites. So after realizing that the non-libgdx TexturePacker could indeed address my problem, I had to determine whether I wanted to push for this feature to be included in libgdx, write it for libgdx by myself, use CodeAndWeb's TexturePacker, or render to mipmappable textures.

I'm leaning toward rendering to intermediate textures because I'd rather not require spritesheet production to be aware of my game's minimum zoom level. This will result in more memory being used per sprite because each *sprite's* dimensions have to be powers of two, but with the spritesheet-based solution, the whole *spritesheet* just needs to have powers-of-two dimensions.

One final note is that I think I was wrong about the `nextPow2` aspect of my last comment; I think it's actually `lastPow2`, e.g. if you are scaling to 1/12 of a sprite's original size, I believe the coordinates in the spritesheet need to be divisible by `lastPow2(12) === 8`.

I'm going to close this issue as I don't think I'm going to need this feature any longer, and if it turns out I do, I'll probably just pay for the non-libgdx TexturePacker.
